pallets/asset-manager/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+pub use pallet::*;
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+pub mod tests;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::{pallet_prelude::*, PalletId};
+	use frame_system::{ensure_root, pallet_prelude::*};
+	use parity_scale_codec::HasCompact;
+	use sp_runtime::traits::AtLeast32BitUnsigned;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	// The registrar trait. We need to comply with this
+	pub trait AssetRegistrar<T: Config> {
+		// How to create an asset
+		fn create_asset(
+			asset: T::AssetId,
+			min_balance: T::Balance,
+			metadata: T::AssetMetaData,
+		) -> DispatchResult;
+	}
+
+	// We implement this trait to be able to get the AssetType and units per second registered
+	impl<T: Config> xcm_primitives::AssetTypeGetter<T::AssetId, T::AssetType> for Pallet<T> {
+		fn get_asset_type(asset_id: T::AssetId) -> Option<T::AssetType> {
+			AssetIdType::<T>::get(asset_id)
+		}
+	}
+
+	impl<T: Config> xcm_primitives::UnitsPerSecondGetter<T::AssetId> for Pallet<T> {
+		fn get_units_per_second(asset_id: T::AssetId) -> Option<u128> {
+			AssetIdUnitsPerSecond::<T>::get(asset_id)
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The Asset Id. This will be used to register the asset in Assets
+		type AssetId: Member + Parameter + Default + Copy + HasCompact + MaxEncodedLen;
+
+		/// The Asset Metadata we want to store
+		type AssetMetaData: Member + Parameter;
+
+		/// The Asset Kind.
+		type AssetType: Parameter + Member + Ord + PartialOrd + Into<Self::AssetId> + Default;
+
+		/// The units in which we record balances.
+		type Balance: Member + Parameter + AtLeast32BitUnsigned + Default + Copy + MaxEncodedLen;
+
+		/// The trait we use to register Assets
+		type AssetRegistrar: AssetRegistrar<Self>;
+
+		/// The AssetManagers's pallet id
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
```

Comment:
what is this used for?

Reply:
nowehere, but when I implement the create asset trait  in the runtime I use this value to set is as the "owner" of the asset. I figured in the future we might use it for, e.g., destroying the asset

Reply:
but we can remove it and add it later. its an associated type, so there is no issue with this

Reply:
Yes, it can be a constant in the runtime instead of an associated type, but If you know we'll need the account in the pallet context later for destroying assets, then it's fine to keep it.

Reply:
@girazoki I found this issue https://github.com/paritytech/substrate/issues/8583 which says `PalletId` should be hardcoded at the runtime level instead of configured at the pallet level -- do you think we should configure this at the runtime ? Your decision, it doesn't make that much of a difference in practice.

Reply:
I think I'm gonna take Shawns advise